stdlib/public/core/UnicodeScalarProperties.swift

```diff
@@ -748,69 +743,68 @@ extension Unicode.Scalar.Properties {
-          Int32(utf16.count),
-          "",
-          &err)
-        guard err.isSuccess else {
-          fatalError("Unexpected error case-converting Unicode scalar.")
+  fileprivate enum _CaseMapping: UInt8 {
+    case uppercase
+    case lowercase
+    case titlecase
+  }
+
+  fileprivate func _getMapping(_ mapping: _CaseMapping) -> String {
+    // First, check if our scalar has a special mapping where it's mapped to
+    // more than 1 scalar.
+    let specialMappingPtr = _swift_stdlib_getSpecialMapping(_scalar.value)
+
+    if let specialMapping = specialMappingPtr {
+      func readSpecialMapping(_ ptr: UnsafePointer<UInt32>) -> String {
+        let count = Int(ptr.pointee)
+
+        if count == 0 {
+          return "\(_scalar)"
         }
-        return Int(correctSize)
-      }
-    }
-    if _fastPath(count <= 16) {
-      fixedArray.count = count
-      return fixedArray.withUnsafeBufferPointer {
-        String._uncheckedFromUTF16($0)
-      }
-    }
-    // Allocate `count` code units on the heap.
-    let array = Array<UInt16>(unsafeUninitializedCapacity: count) {
-      buf, initializedCount in
-      _scalar.withUTF16CodeUnits { utf16 in
-        var err = __swift_stdlib_U_ZERO_ERROR
-        let correctSize = u_strTo(
-          buf.baseAddress._unsafelyUnwrappedUnchecked,
-          Int32(buf.count),
-          utf16.baseAddress._unsafelyUnwrappedUnchecked,
-          Int32(utf16.count),
-          "",
-          &err)
-        guard err.isSuccess else {
-          fatalError("Unexpected error case-converting Unicode scalar.")
+
+        var result = ""
+
+        for i in 0 ..< count {
+          result += "\(Unicode.Scalar(_unchecked: ptr[1 + i]))"
         }
-        _internalInvariant(count == correctSize, "inconsistent ICU behavior")
-        initializedCount = count
+
+        return result
+      }
+
+      switch mapping {
+      case .uppercase:
+        return readSpecialMapping(specialMapping)
+
+      case .lowercase:
+        let upperCount = Int(specialMapping.pointee)
+
+        return readSpecialMapping(specialMapping + upperCount + 1)
+
+      case .titlecase:
+        let upperCount = Int(specialMapping.pointee)
+        let lowerPtr = specialMapping + upperCount + 1
+        let lowerCount = Int(lowerPtr.pointee)
+
+        return readSpecialMapping(lowerPtr + lowerCount + 1)
       }
     }
-    return array.withUnsafeBufferPointer {
-      String._uncheckedFromUTF16($0)
+
+    // If we did not have a special mapping, check if we have a direct scalar
+    // to scalar mapping.
+    let mappingDistance = _swift_stdlib_getMapping(
+      _scalar.value,
+      mapping.rawValue
+    )
+
+    if mappingDistance != 0 {
+      let scalar = Unicode.Scalar(
+        _unchecked: UInt32(Int(_scalar.value) &+ Int(mappingDistance))
+      )
+      return "\(scalar)"
     }
+
+    // We did not have any mapping. Return the scalar as is.
+    return "\(_scalar)"
```

Comment:
I'd recommend using `String(_scalar)` instead.

If you look at the [generated code](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:swift,selection:(endColumn:19,endLineNumber:6,positionColumn:19,positionLineNumber:6,selectionStartColumn:19,selectionStartLineNumber:6,startColumn:19,startLineNumber:6),source:'func+scalarToStr_1(_+scalar:+Unicode.Scalar)+-%3E+String+%7B%0A++++%22%5C(scalar)%22%0A%7D%0A%0Afunc+scalarToStr_2(_+scalar:+Unicode.Scalar)+-%3E+String+%7B%0A++++String(scalar)%0A%7D'),l:'5',n:'0',o:'Swift+source+%231',t:'0')),k:34.20365535248042,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:swiftnightly,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:swift,libs:!(),options:'-O',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'x86-64+swiftc+nightly+(Swift,+Editor+%231,+Compiler+%231)',t:'0')),k:65.79634464751958,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4), using a string interpolation does a bunch of extra work, creating a string and appending to it via `Swift.Unicode.Scalar.write<A where A: Swift.TextOutputStream>(to: inout A) -> ()` (which just [does the same thing](https://github.com/apple/swift/blob/85d9507fde3deb5889f71f81f09a05319898b029/stdlib/public/core/OutputStream.swift#L616) anyway).

The initialiser is more direct, and cuts out a bunch of unnecessary work.